keps/0019-package-api-versioning.md

```diff
@@ -0,0 +1,107 @@
+  2. A user should be able to install a specific `appVersion` of an operator
+
+### Support multiple API versions in the package repository
+
+Through the `apiVersion` field in an `operator.yaml` it is already guaranteed that the operator will match a specific API version of KUDO. However, this field isn't taken into account when installing packages from a repository. KUDO will try to install the latest version of the package even if the `apiVersion` doesn't match. This can be mitigated by adding the packages `apiVersion` to the repository index and filtering out all packages that don't have a matching version. This way, older versions of packages can still be part of the repository and keep working with older versions of KUDO.
+
+For this, the following changes have to be made in KUDO:
+
+* Add `apiVersion` to the package index by updating the `Metadata` struct in `pkg/kudoctl/util/repo` and filling these values in `pkg/kudoctl/util/repo.ToPackageVersion`. A `index.yaml` entry should then look like this:
+
+```
+  kafka:
+  - appVersion: 2.3.0
+    digest: e80c7b783d327190d489159e89e0a005a6a8b00610bdb7e8b1bea73c49bf485a
+    maintainers:
+    - email: zmalikshxil@gmail.com
+      name: Zain Malik
+    name: kafka
+    urls:
+    - https://kudo-repository.storage.googleapis.com/kafka-0.2.0.tgz
+    version: 0.2.0
+    apiVersion: kudo.dev/v1beta1
+```
+
+* Update `pkg/kudoctl/util/repo.ParseIndexFile` to filter packages by API version. It should filter out packages whose `apiVersion` doesn't match the current `apiVersion` of KUDO.
+
+### Install a specific application version
+
+As `appVersion` is already part of the repository index, a command line flag `--app-version` will be added to `kubectl kudo install` to select a specific application version. Because this version is application specific and might not follow [semantic versioning](https://semver.org/), the absence of this command line flag will result in KUDO installing the latest `version` of a package, or the version indicated with the `--version` flag.
+
+### Naming of package tarballs
+
+To avoid disambiguities, the `appVersion` of a package will be part of the tarball name. E.g., the Kafka package described above will be named `kafka-2.3.0-0.2.0.tgz`. For this, changes in `kudoctl/packages/writer` are necessary.
+
+### Semantic versioning
+
+`kudoVersion` and `version` follow the [Semantic Versioning Specification](https://semver.org/). `appVersion` is application dependent and might not follow this specification. It should be treated as a free-form string without ordering. The following filtering should be used when installing packages from a repository:
+  1. Filter by `apiVersion` of packages that match the one used by `kubectl kudo`
+  2. Filter by `kudoVersion` of packages that are smaller or equal than the one of `kubectl kudo`
+  3. Filter by `appVersion` of package if provided by user
+  4. Use latest or user-provided `version` of package
+
```

Comment:
I'm not sure if it goes in this kep or another... but I believe we should have a way for an operator to define required features in k8s.   today there are huge assumptions regarding k8s version and features... what we should do is have something like GVK defined in the `operator.yaml`.  then we are left with deciding if we search for an operator based on this... or fail if the "latest" filtered operator has requirements which are not met.

Reply:
agree with @kensipe here. And also not sure it would be part of this KEP, but this would be a big nice to have feature.

Reply:
Yes, this is needed. But I'd rather have a separate KEP for this. Especially the Kubernetes feature handling sounds like quite some additional code will be needed in KUDO to handle this.